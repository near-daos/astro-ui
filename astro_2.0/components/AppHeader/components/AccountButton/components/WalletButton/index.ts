export * from 'astro_2.0/components/AppHeader/components/AccountButton/components/WalletButton/WalletButton';
