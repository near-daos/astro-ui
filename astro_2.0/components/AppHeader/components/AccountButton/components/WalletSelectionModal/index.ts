export * from 'astro_2.0/components/AppHeader/components/AccountButton/components/WalletSelectionModal/WalletSelectionModal';
