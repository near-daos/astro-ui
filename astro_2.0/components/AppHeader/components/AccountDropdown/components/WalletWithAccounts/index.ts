export * from 'astro_2.0/components/AppHeader/components/AccountDropdown/components/WalletWithAccounts/WalletWithAccounts';
