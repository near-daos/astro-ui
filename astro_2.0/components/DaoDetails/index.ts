export * from './DaoDetailsMinimized';
export * from './DaoDetailsGrid';
