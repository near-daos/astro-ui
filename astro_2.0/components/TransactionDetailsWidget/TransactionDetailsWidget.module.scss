@use 'styles/breakpoints';

@import 'styles/mixins';
@import 'styles/typography';

.root {
  align-items: center;
  background: var(--color-neutral-10);
  border-bottom: 0.5px solid var(--color-brand-purple);
  border-left: 0.5px solid var(--color-brand-purple);
  border-radius: 0 0 7px 7px;
  border-right: 0.5px solid var(--color-brand-purple);
  box-shadow: 3px 2px 24px rgba(123, 97, 255, 0.25);
  color: var(--color-neutral-10);
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  min-height: 80px;
  padding: 12px;

  @include tablet-portrait {
    flex-direction: row;
  }

  @include tablet-landscape {
    padding: 0 28px;
  }

  @include astro-desktop {
    align-items: flex-start;
    flex-direction: column;
  }

  @include large-desktop {
    align-items: center;
    flex-direction: row;
  }
}

.topBorder {
  border-radius: 7px;
  border-top: 0.5px solid var(--color-brand-purple);
}

.mainContent {
  align-items: center;
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
  width: 100%;

  @include tablet-landscape {
    flex-direction: row;
  }
}

.transactionDetails {
  align-items: flex-end;
  display: flex;
  flex-basis: 46%;
  justify-content: space-between;
  margin-bottom: 12px;
  padding: 0 24px;
  width: 100%;

  @include tablet-landscape {
    justify-content: flex-start;
    margin-bottom: 0;
    padding: 0;
  }

  .detailsItem {
    @include tablet-landscape {
      margin-left: 32px;
    }
  }

  .tokenLabel {
    margin-left: 4px;
  }
}

.infoBlock {
  margin-top: 8px;

  &:first-of-type {
    margin-right: 32px;
  }
}

.warningContainer {
  margin: 8px 0;
}

.warning {
  color: var(--color-error-60);
}

.createDao {
  margin: 8px 0;
}

.row {
  align-items: center;
  color: var(--color-black);
  display: flex;
  white-space: nowrap;

  .iconWrapper {
    border-radius: 50%;
    height: 24px;
    margin-right: 8px;
    width: 24px;
  }
}

.inputWrapper {
  /* stylelint-disable */
  /* Chrome, Safari, Edge, Opera */
  input::-webkit-outer-spin-button,
  input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  /* Firefox */
  input[type='number'] {
    -moz-appearance: textfield;
  }
  /* stylelint-enable */

  input {
    background-color: var(--color-neutral-10);
    font-weight: 700;
    height: 24px !important;
    padding: 0 !important;
    white-space: nowrap;
    word-break: normal;
    @include placeholder {
      color: var(--color-neutral-40);
      transition: opacity 0.2s linear;
    }
  }
}

.warningContainer ~ .mainContent {
  @include tablet-landscape {
    justify-content: flex-end;
  }
}

.warningContainer ~ .mainContent .transactionDetails {
  @include astro-mobile {
    justify-content: space-around;
    padding: 0;
  }

  @include astro-desktop {
    justify-content: flex-end;
  }
}

.warningContainer ~ .mainContent .detailsItem {
  @include tablet-landscape {
    margin: 0 32px 0 0;
  }
}
