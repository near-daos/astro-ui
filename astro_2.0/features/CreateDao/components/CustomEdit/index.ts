import CustomEdit from './CustomEdit';

export default CustomEdit;
