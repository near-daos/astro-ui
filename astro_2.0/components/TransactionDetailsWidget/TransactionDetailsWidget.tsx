import React from 'react';
import cn from 'classnames';
import { useFormContext } from 'react-hook-form';

import { Button } from 'components/button/Button';
import { InfoBlockWidget } from 'astro_2.0/components/InfoBlockWidget';
import { Input } from 'components/inputs/input/Input';
import { InputWrapper } from 'astro_2.0/features/CreateProposal/components/InputWrapper';
import { formatYoktoValue } from 'helpers/format';
import { DEFAULT_PROPOSAL_GAS } from 'services/sputnik/constants';

import styles from './TransactionDetailsWidget.module.scss';

interface InfoWidgetProps {
  label?: string;
  value: string;
}

interface CreateProposalWidgetProps {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onSubmit: (data: any) => Promise<void>;
  bond: InfoWidgetProps;
  gas?: InfoWidgetProps;
  warning?: string;
  buttonLabel?: string;
  standAloneMode?: boolean;
}

export const TransactionDetailsWidget: React.FC<CreateProposalWidgetProps> = ({
  onSubmit,
  bond,
  gas,
  standAloneMode = false,
  buttonLabel = 'Purpose',
  warning,
}) => {
  const { handleSubmit, register } = useFormContext();

  function renderWarning() {
    if (warning) {
      return (
        <InfoBlockWidget
          label="Warning"
          value={warning}
          className={styles.warningContainer}
          valueClassName={styles.warning}
        />
      );
    }

    return null;
  }

  const gasInputProps = gas
    ? {
        readOnly: true,
        defaultValue: gas.value,
      }
    : {
        placeholder: `${DEFAULT_PROPOSAL_GAS}`,
        ...register('gas'),
      };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className={cn(styles.root, { [styles.topBorder]: standAloneMode })}>
        {renderWarning()}
        <div className={styles.mainContent}>
          <div className={styles.transactionDetails}>
            <InputWrapper fieldName="gas" label="Gas">
              <div className={styles.row}>
                <Input
                  className={cn(styles.inputWrapper, styles.narrow)}
                  type="number"
                  min={0.01}
                  step={0.01}
                  max={0.25}
                  isBorderless
                  size="block"
                  {...gasInputProps}
                />
                <div>NEAR</div>
              </div>
            </InputWrapper>

            <InputWrapper fieldName="bond" label="Bond">
              <div className={styles.row}>
                <Input
                  className={cn(styles.inputWrapper, styles.narrow)}
                  readOnly
                  type="number"
                  min={0.01}
                  step={0.01}
                  max={0.25}
                  placeholder="0.25"
                  defaultValue={formatYoktoValue(bond.value)}
                  isBorderless
                  size="block"
                />
                <div>NEAR</div>
              </div>
            </InputWrapper>
          </div>
          <Button
            type="submit"
            size="medium"
            variant="black"
            className={styles.createDao}
          >
            {buttonLabel}
          </Button>
        </div>
      </div>
    </form>
  );
};
