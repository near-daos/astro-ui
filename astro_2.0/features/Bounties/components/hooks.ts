import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
