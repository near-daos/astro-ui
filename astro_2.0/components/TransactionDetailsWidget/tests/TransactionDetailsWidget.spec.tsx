/* eslint-disable @typescript-eslint/ban-ts-comment, @typescript-eslint/no-empty-function */
import { render } from 'jest/testUtils';
import { useFormContext } from 'react-hook-form';

import { TransactionDetailsWidget } from 'astro_2.0/components/TransactionDetailsWidget';

jest.mock('react-hook-form', () => {
  return {
    useFormContext: jest.fn(() => ({
      watch: () => 1,
      register: () => ({}),
      handleSubmit: (callback: () => void) => callback(),
      formState: {
        errors: {},
      },
    })),
  };
});

describe('Transaction details widget', () => {
  it('Should render component', () => {
    const { container } = render(
      <TransactionDetailsWidget
        onSubmit={async () => {}}
        bond={{ value: '2' }}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render warning', () => {
    const warning = 'This is my warning';

    const { getAllByText } = render(
      <TransactionDetailsWidget
        onSubmit={async () => {}}
        bond={{ value: '2' }}
        warning={warning}
      />
    );

    expect(getAllByText(warning)).toHaveLength(1);
  });

  it.each`
    gas              | width
    ${'1000000'}     | ${'7ch'}
    ${'10000000000'} | ${'10ch'}
  `(
    'Should render input of proper width depending on amount of gas',
    ({ gas, width }) => {
      // @ts-ignore
      useFormContext.mockImplementation(() => ({
        watch: () => gas,
        register: () => ({}),
        handleSubmit: (callback: () => void) => callback(),
        formState: {
          errors: {},
        },
      }));

      const { getByTestId } = render(
        <TransactionDetailsWidget
          onSubmit={async () => {}}
          bond={{ value: '2' }}
        />
      );

      const input = getByTestId('gas-input');

      expect(input.style.width).toEqual(width);
    }
  );
});
