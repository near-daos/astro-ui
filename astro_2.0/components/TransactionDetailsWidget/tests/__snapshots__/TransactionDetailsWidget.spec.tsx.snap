// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction details widget Should render component 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class=""
    >
      <div>
        <div>
          <div
            class=""
          >
            <div
              class=""
            >
              <span>
                components.transactionDetailsWidget.tGas
              </span>
            </div>
            <div>
              <label
                class="undefined"
                for="1"
              >
                <input
                  class=""
                  data-testid="gas-input"
                  id="1"
                  max="300"
                  min="10"
                  placeholder="150"
                  step="1"
                  style="text-align: left; width: 6ch;"
                  type="number"
                />
              </label>
            </div>
          </div>
          <div
            class=""
          >
            <div
              class=""
            >
              <span>
                components.transactionDetailsWidget.bond
              </span>
            </div>
            <div>
              <label
                class="undefined"
                for="2"
              >
                <input
                  class=""
                  id="2"
                  max="0.25"
                  min="0.01"
                  placeholder="0.25"
                  readonly=""
                  step="0.01"
                  style="text-align: left; width: 5ch;"
                  type="number"
                  value="0"
                />
              </label>
              <div>
                NEAR
              </div>
            </div>
          </div>
        </div>
        <button
          class=""
          type="submit"
        >
          Purpose
        </button>
      </div>
    </div>
  </form>
</div>
`;
